src/main/java/htsjdk/utils/ClassFinder.java

```diff
@@ -0,0 +1,192 @@
+                } catch (URISyntaxException e) {
+                    log.warn("Cannot convert to URI the " + urlPath + " URL");
+                }
+                if (urlPath.indexOf('!') > 0) {
+                    urlPath = urlPath.substring(0, urlPath.indexOf('!'));
+                }
+                if (jarPath != null && !jarPath.equals(urlPath)) {
+                    continue;
+                }
+
+                //Log.info("Looking for classes in location: " + urlPath);
+                final File file = new File(urlPath);
+                if ( file.isDirectory() ) {
+                    scanDir(file, packageName);
+                }
+                else {
+                    scanJar(file, packageName);
+                }
+            }
+            catch (IOException ioe) {
+                log.warn("could not read entries", ioe);
+            }
+        }
+    }
+
+    /**
+     * Scans the entries in a ZIP/JAR file for classes under the parent package.
+     * @param file the jar file to be scanned
+     * @param packagePath the top level package to start from
+     */
+    protected void scanJar(final File file, final String packagePath) throws IOException {
+        final ZipFile zip = new ZipFile(file);
+        final Enumeration<? extends ZipEntry> entries = zip.entries();
+        while ( entries.hasMoreElements() ) {
+            final ZipEntry entry = entries.nextElement();
+            final String name = entry.getName();
+            if (name.startsWith(packagePath)) {
+                handleItem(name);
+            }
+        }
+    }
+
+    /**
+     * Scans a directory on the filesystem for classes.
+     * @param file the directory or file to examine
+     * @param path the package path acculmulated so far (e.g. edu/mit/broad)
+     */
+    protected void scanDir(final File file, final String path) {
+        for ( final File child: file.listFiles() ) {
+            final String newPath = (path==null ? child.getName() : path + '/' + child.getName() );
+            if ( child.isDirectory() ) {
+                scanDir(child, newPath);
+            }
+            else {
+                handleItem(newPath);
+            }
+        }
+    }
+
+    /**
+     * Checks an item to see if it is a class and is annotated with the specified
+     * annotation.  If so, adds it to the set, otherwise ignores it.
+     * @param name the path equivelant to the package + class/item name
+     */
+    protected void handleItem(final String name) {
+        if (name.endsWith(".class")) {
+            final String classname = toClassName(name);
+
+            try {
+                final Class<?> type = loader.loadClass(classname);
+                if (parentType.isAssignableFrom(type)) {
+                    this.classes.add(type);
+                }
+            }
+            catch (Throwable t) {
+                log.debug("could not load class: " + classname, t);
+            }
+        }
+    }
+
```

Comment:
Can you add the following additional methods to this?  Then we can delete our copy in gatk.

```
    /**
     * Fetches the set of classes discovered so far, subsetted down to concrete (non-abstract/interface) classes only
     *
     * @return subset of classes discovered so far including only concrete (non-abstract/interface) classes
     */
    public Set<Class<?>> getConcreteClasses() {
        final Set<Class<?>> concreteClassSet = new LinkedHashSet<>();

        for ( final Class<?> clazz : classes ) {
            if ( isConcrete(clazz) ) {
                concreteClassSet.add(clazz);
            }
        }

        return concreteClassSet;
    }

    /**
     * Determines whether or not the specified class is concrete (ie., non-abstract and non-interface)
     *
     * @param clazz class to check
     * @return true if the class is neither abstract nor an interface, otherwise false
     */
    public static boolean isConcrete( final Class<?> clazz ) {
        return ! Modifier.isAbstract(clazz.getModifiers()) && ! Modifier.isInterface(clazz.getModifiers());
    }
```

Reply:
done.